File: java/src/test/java/com/bc/ur/URDecoderTest.java
package com.bc.ur;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.Arrays;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;


@RunWith(JUnit4.class)
public class URDecoderTest {

    @Test
    public void testDecodeSinglePart() {
        UR expectedUR = UR.create(50, "Wolf");
        String encoded =
                "ur:bytes/hdeymejtswhhylkepmykhhtsytsnoyoyaxaedsuttydmmhhpktpmsrjtgwdpfnsboxgwlbaawzuefywkdplrsrjynbvygabwjldapfcsdwkbrkch";
        UR ur = URDecoder.decode(encoded);
        assertEquals(expectedUR.getType(), ur.getType());
        assertTrue(Arrays.deepEquals(TestUtils.toTypedArray(expectedUR.getCbor()), TestUtils.toTypedArray(ur.getCbor())));
    }

    @Test
    public void testDecodeMultiParts() {
        UR ur = UR.create(32767, "Wolf");
        UREncoder encoder = new UREncoder(ur, 1000, 100, 10);
        URDecoder decoder = new URDecoder();
        do {
            String part = encoder.nextPart();
            decoder.receivePart(part);
        } while (!decoder.isComplete());

        if (decoder.isSuccess()) {
            UR resultUR = decoder.resultUR();
            assertEquals(ur.getType(), resultUR.getType());
            assertTrue(Arrays.deepEquals(TestUtils.toTypedArray(ur.getCbor()), TestUtils.toTypedArray(resultUR.getCbor())));
        } else {
            throw decoder.resultError();
        }
    }

    @Test(expected = IllegalStateException.class)
    public void testDecodeError() {
        String[] invalidData = new String[]{
                "",
                "ur:bytes/",
                "ur:ur:ur",
                "uf:bytes/hdeymejtswhhylkepmykhhtsytsnoyoyaxaedsuttydmmhhpktpmsrjtgwdpfnsboxgwlbaawzuefywkdplrsrjynbvygabwjldapfcsdwkbrkch"
        };
        for (String it : invalidData) {
            URDecoder.decode(it);
        }
    }
}

File: android/app/src/androidTest/java/com/bc/ur/URDecoderTest.kt
package com.bc.ur

import androidx.test.ext.junit.runners.AndroidJUnit4
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import java.util.*


@RunWith(AndroidJUnit4::class)
class URDecoderTest {

    @Test
    fun testDecodeSinglePart() {
        val expectedUR = UR.create(50, "Wolf")
        val encoded =
            "ur:bytes/hdeymejtswhhylkepmykhhtsytsnoyoyaxaedsuttydmmhhpktpmsrjtgwdpfnsboxgwlbaawzuefywkdplrsrjynbvygabwjldapfcsdwkbrkch"
        val ur = URDecoder.decode(encoded)
        assertEquals(expectedUR.type, ur.type)
        assertTrue(expectedUR.cbor.toTypedArray().contentDeepEquals(ur.cbor.toTypedArray()))
    }

    @Test
    fun testDecodeMultiParts() {
        val ur = UR.create(32767, "Wolf")
        val encoder = UREncoder(ur, 1000, 100, 10)
        val decoder = URDecoder()
        do {
            val part = encoder.nextPart()
            decoder.receivePart(part)
        } while (!decoder.isComplete)

        if (decoder.isSuccess) {
            val resultUR = decoder.resultUR()
            assertEquals(ur.type, resultUR.type)
            assertTrue(ur.cbor.toTypedArray().contentDeepEquals(resultUR.cbor.toTypedArray()))
        } else {
            val ex = decoder.resultError()
            throw ex
        }
    }

    @Test(expected = IllegalStateException::class)
    fun testDecodeError() {
        arrayOf(
            "",
            "ur:bytes/",
            "ur:ur:ur",
            "uf:bytes/hdeymejtswhhylkepmykhhtsytsnoyoyaxaedsuttydmmhhpktpmsrjtgwdpfnsboxgwlbaawzuefywkdplrsrjynbvygabwjldapfcsdwkbrkch"
        ).forEach {
            URDecoder.decode(it)
        }

    }
}